### ISSUE TITLE
degraded intellisense/error reporting for launch.json

### ISSUE THREAD

**Issue by weinand** [microsoft/vscode#142642]
A side effect of the schema optimisation work done via https://github.com/microsoft/vscode/issues/120711 is a degradation of intellisense and error reporting for launch.json.

In particular the following areas are affected:
- unknown properties are not flagged as errors,
- preLaunchTask: when using a structured preLaunchTask (instead of just the name of a task), there is no intellisense/error reporting for the task attributes (IIRC structured preLaunchTasks should only allow something like: `{"task": "...", "type": "..." }` ),
- there might be more...

**Comment by roblourens** [comment]
The first issue is a missing "additionalProperties=false". The second, I can't figure out. In OSS, I get the correct intellisense and errors for preLaunchTask. In insiders with no extensions/fresh user data dir, I don't. But when I copy the full schema from inside vscode into a json file and reference it with `"$schema"` in the launch.json, then it works. @aeschli any idea what would be going on there?

**Comment by roblourens** [comment]
For the first part I will have to restructure the allOf section and redeclare properties, as in https://json-schema.org/understanding-json-schema/reference/object.html#extending-closed-schemas. But there is a new property that would help, which our JSON validation hasn't implemented: https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties

**Comment by weinand** [comment]
@aeschli any chance that VS Code will support `unevaluatedProperties`?

**Comment by aeschli** [comment]
I can look at it for May: https://github.com/microsoft/vscode-json-languageservice/issues/132

**Comment by weinand** [comment]
That would be awesome!

**Comment by roblourens** [comment]
Even without `unevaluatedProperties`, from my understanding, I only have to add the 3 OS properties to each config. They still use a $ref so it's not redeclaring every single property. 

The one I can't figure out is why the preLaunchTask schema isn't working in Insiders (with no extensions) but is in OSS. I pull in the task schema here: https://github.com/microsoft/vscode/blob/4a130c40ed876644ed8af2943809d08221375408/src/vs/workbench/contrib/debug/browser/debugAdapterManager.ts#L132

**Comment by weinand** [comment]
@aeschli any idea why a schema works in OSS but not in Insiders?

**Comment by aeschli** [comment]
@weinand No idea. How can I reproduce?

**Comment by weinand** [comment]
@roblourens could you please add steps?

**Comment by roblourens** [comment]
> preLaunchTask: when using a structured preLaunchTask (instead of just the name of a task), there is no intellisense/error reporting for the task attributes (IIRC structured preLaunchTasks should only allow something like: {"task": "...", "type": "..." } ),

- Add `"preLaunchTask": { }`, trigger intellisense inside, I get no suggestions.
- Add `"preLaunchTask": { "blah": 1 }`, I don't get an error about the unknown property

**Comment by aeschli** [comment]
Here's what I see in the latest insiders: (1.67.0-insider, commit: 6f26fa19db7e4cd2af57f52fd7009e715c3d7aab)

- open `launch.json` in the vscode repo on main
- In the first configuration add a `preLaunchTask`

![image](https://user-images.githubusercontent.com/6461412/165247697-231d03ab-7ff2-4a9a-afca-9cf5c486fe4c.png)

![image](https://user-images.githubusercontent.com/6461412/165247897-2c56064f-d433-4d51-a178-a1e280975499.png)

- Also when running from sources:

![image](https://user-images.githubusercontent.com/6461412/165292409-b03e9013-d514-4f21-98f8-b63333d6110a.png)


I think what's special here is that preLaunchTask is a `oneOf` many variants. Each variant has a different `type` (`gulp`, `grunt`, ...). Until `type` is defined, it feels random which variant is picked by the JSON language service and which properties are proposed.

For the best user experience, IMO 'preLaunchTask` has default snippets for each type. 
So instead of
![image](https://user-images.githubusercontent.com/6461412/165293730-17950aec-57a1-44b3-9842-89b4af6d0188.png)

there's a proposal for each type. That way the user select4s a type early and the completions make sense

**Comment by roblourens2** [comment]
Weird, still doesn't work for me:
![Recording 2022-04-27 at 10 26 42](https://user-images.githubusercontent.com/64314422/165584693-f0c5b138-2f6a-4aff-bc51-29e9f1bd18a9.gif)

This is in a totally clean insiders window, fresh user data dir, empty extensions dir. No completions at all inside preLaunchTask.

**Comment by aeschli** [comment]
In the status line, can you click on the `{}` next to JSON, select `Show Scheams` and open the schema.
Copy the schema to a new empty editor. Maybe a syntax error in the schema?

**Comment by roblourens2** [comment]
I didn't know about that - have been copying the schema out of the debugger which is hard. Ended up with this somehow:

```json
"preLaunchTask": {
	"anyOf": [{
		"oneOf": []
	}, {
		"type": ["string"]
	}],
	"default": "",
	"defaultSnippets": [{
		"body": {
			"task": "",
			"type": ""
		}
	}],
	"description": "Task to run before debug session starts."
},
```

so I think that this `const taskSchema = TaskDefinitionRegistry.getJsonSchema();` returns an empty `oneOf`.

I notice that it caches the result, so maybe in my Insiders it's being called before the task types are registered. If that list of contributions changes, the launch schema won't get the new options. @alexr00

**Comment by alexr00** [comment]
@roblourens I tried this out and I'm seeing 10 elements in the `oneOf`. Does this happen every time for you?

**Comment by roblourens** [comment]
Yes, but I think it depends on how early `getJsonSchema` is first called. I have a lot of extensions installed, something must be different.

**Comment by aeschli** [comment]
`unevaluatedProperties` is now supported by the JSON language service.

**Comment by weinand** [comment]
@aeschli great, thanks a lot!

**Comment by alexr00** [comment]
@roblourens I'm not sure I understand what is needed from the tasks side. Should we add an event that fires when the schema changes?

**Comment by roblourens** [comment]
I'm pretty sure that getJsonSchema is caching an incomplete value and returning that even after more registrations have happened https://github.com/microsoft/vscode/blob/1b88f28b495c90bb656bc3fb710a49d662074a3e/src/vs/workbench/contrib/tasks/common/taskDefinitionRegistry.ts#L155

I need that to return the real latest value. There is already an event for change although I'm not listening to it right now
